[Docs] Add docs for custom entity with access per admin channel (#12619)
Sylius/Sylius: docs/cookbook/entities/custom-model-accessible-per-admin-channel.rst

        {
            $listQueryBuilder = $this->supplierRepository->getQueryBuilder();

            /** @var ChannelInterface|null $channel */
            $channel = $this->adminChannelProvider->getChannel();
            if ($channel === null) {
                return $listQueryBuilder;
            }

            return $listQueryBuilder
                ->andWhere('o.channel = :channel')
                ->setParameter('channel', $channel)
            ;
        }
    }

.. code-block:: yaml

    App\Doctrine\ORM\RestrictingSupplierListQueryBuilder:
        public: true
        class: App\Doctrine\ORM\RestrictingSupplierListQueryBuilder
        arguments: ['@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface', '@app.repository.supplier']

* Add method to Suppliers grid:

.. code-block:: yaml

    sylius_grid:
        grids:
            app_admin_supplier:
                driver:
                    name: doctrine/orm
                    options:
                        class: App\Entity\Supplier
                        repository:
  +                         method: [expr:service('App\\Doctrine\\ORM\\RestrictingSupplierListQueryBuilder'), create]

>   Well done! That's it, now you have a Supplier entity, that is accessible within the Sylius Plus Administrators per Channel feature!

Reviewer: ```suggestion
Well done! That's it, now you have a Supplier entity, that is accessible within the Sylius Plus Administrators per Channel feature!
```

.. rst-class:: plus-doc

> How to add a custom model accessible per admin channel?

Reviewer: ```suggestion
How to add a custom model accessible for respective channel administrators?
```

=======================================================

> In some cases you may be needing to add new model to your application that is accessible on admin site per admin channel

Reviewer: ```suggestion
Given that you are using Sylius Plus, the licensed edition of Sylius, you may have the Administrators per Channel defined in your application. Thus when you add a new, channel-based entity to it, you will need to enable this entity to be viewed only by the relevant channel admins.
```

> 1. Define custom model, our example will be basic on **Supplier entity**

Reviewer: ```suggestion
1. Define your custom model, our example will be the **Supplier entity**
```

------------------------------------------------------------------------

> For making entity use guide: :doc:`here </cookbook/entities/custom-model>`.

Reviewer: ```suggestion
In order to prepare a simple Entity follow this guide: :doc:`here </cookbook/entities/custom-model>`.
```

> For making entity use guide: :doc:`here </cookbook/entities/custom-model>`.

Reviewer 2: And maybe `creating an` would be better than `making`?

> Next create channel field with relation to ``channel`` entity:

Reviewer: ```suggestion
* Having your Supplierentity created, add a channel field with relation to the ``Channel`` entity:
```

.. code-block:: php

    /**
      * @var ChannelInterface
      * @ORM\ManyToOne(targetEntity="Sylius\Plus\Entity\ChannelInterface")
      * @ORM\JoinColumn(name="channel_id", referencedColumnName="id", nullable=true)
      */
     protected $channel;

Create properly migration for this changes

> In order to be owned access to choosing only proper channel, create ChannelPerAdminChannelType:

Reviewer 3: ```suggestion
In order to have access to choosing only proper channel, create ChannelPerAdminChannelType:
```

Reviewer 2: ```suggestion
In order to have access for choosing only the proper channel, create ChannelPerAdminChannelType:
```
🤔

> In order to be owned access to choosing only proper channel, create ChannelPerAdminChannelType:

Reviewer: ```suggestion
1. Adjust the entity's form type to also be channel-based:
---------------------------------------------------------

In order to have access to choosing a proper channel, create ``ChannelPerAdminChannelType``:
```

Reviewer: So shouldn't it be "ChannelsOfAdminChannelType"?

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Form;

    use Sylius\Component\Resource\Repository\RepositoryInterface;
    use Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface;
    use Sylius\Plus\Entity\ChannelInterface;
    use Symfony\Bridge\Doctrine\Form\DataTransformer\CollectionToArrayTransformer;
    use Symfony\Component\Form\AbstractType;
    use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
    use Symfony\Component\Form\FormBuilderInterface;
    use Symfony\Component\OptionsResolver\Options;
    use Symfony\Component\OptionsResolver\OptionsResolver;

>     final class ChannelPerAdminChannelType extends AbstractType

Reviewer 3: Can we have this form type predefined in Plus and use it somewhere?

Author: Yep. I think we should replace all places with `ChannelChoiceType` for `ChannelPerAdminChannelChoiceType`. This is not scope this task, but good improvement in future. Next good improvement is to add resources name as parameters in services, without decorating `ResourceChannelEnabilibityChecker` service:
Now:
```
final class ResourceChannelEnabilibityChecker implements ResourceChannelEnabilibityCheckerInterface
{
    public function forResourceName(string $resourceName): bool
    {
        return in_array($resourceName, [
            'credit_memo',
            'customer',
            'invoice',
            'order',
            'payment',
            'product',
            'product_variant',
            'return_request',
            'shipment',
        ]);
    }
}
```

It could be more generic, for example:
```
final class ResourceChannelEnabilibityChecker implements ResourceChannelEnabilibityCheckerInterface
{
    private $resources;

    public function __construct($resources)
    {
        $this->resources = array_merge(
            $resources,
            [
                'credit_memo',
                'customer',
                'invoice',
                'order',
                'payment',
                'product',
                'product_variant',
                'return_request',
                'shipment',
            ]
        );
    }

    public function forResourceName(string $resourceName): bool
    {
        return in_array($resourceName, $this->resources);
    }
```
After this changes, final application needs overwrite only service and add new entity names

            }
        }

        public function configureOptions(OptionsResolver $resolver): void
        {
            $resolver->setDefaults([
                'choices' => function (Options $options): array {
                    /** @var ChannelInterface|null $adminChannel */
                    $adminChannel = $this->adminChannelProvider->getChannel();

                    if ($adminChannel != null) {
                        return [$adminChannel];
                    }

                    return $this->channelRepository->findAll();
                },
                'choice_value' => 'code',
                'choice_label' => 'name',
                'choice_translation_domain' => false,
            ]);
        }

        public function getParent(): string
        {
            return ChoiceType::class;
        }

        public function getBlockPrefix(): string
        {
            return 'sylius_channel_choice';
        }
    }

.. code-block:: yaml

    App\Form\ChannelPerAdminChannelType:
        arguments: ['@sylius.repository.channel', '@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface']
        tags: ['form.type']

> Service ``Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface`` allows to get proper channel for present admin.

Reviewer: ```suggestion
The ``Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface`` service allows getting a list of proper channels for the currently logged-in admin.
```

                model: App\Entity\Supplier
                repository: App\Repository\SupplierRepository
   +            form: App\Form\SupplierType

1. Add your entity to admin menu:
----------------------------------

Adding new entity to admin menu is described in point ``How to customize Admin Menu`` into :doc:`this </customization/menu>`. guide

1. Add access via administrator roles (ACL/RBAC):
-------------------------------------------------

Add access to all resource sections: ``index, create, update, delete, show, bulk_delete``

.. code-block:: yaml

    sylius_plus:
    permissions:
        app_admin_supplier_index:
            parent: suppliers
            label: action.index
        app_admin_supplier_create:
            parent: suppliers
            label: action.create
        app_admin_supplier_update:
            parent: suppliers
            label: action.update
        app_admin_supplier_delete:
            parent: suppliers
            label: action.delete
        app_admin_supplier_show:
            parent: suppliers
            label: action.show
        app_admin_supplier_bulk_delete:
            parent: suppliers
            label: action.bulk_delete

> More information about accessing administrator roles (ACL/RBAC) you can find :doc:`here </book/customers/admin_user>`.

Reviewer: `.. note::` then :)

    {
        /** @var ResourceChannelCheckerInterface */
        private $decoratedResourceChannelChecker;

        public function __construct(ResourceChannelCheckerInterface $decoratedResourceChannelChecker)
        {
            $this->decoratedResourceChannelChecker = $decoratedResourceChannelChecker;
        }

        public function isFromChannel(object $resource, ChannelInterface $channel): bool
        {
            // this condition is needed because while creating symfony form(our edit/create section), given object has all fields as null. After created an empty form this iteration is repeat and properly checked
            if ($resource instanceof Supplier && $resource->getChannel() === null) {
                return true;
            }

            return $this->decoratedResourceChannelChecker->isFromChannel($resource, $channel);
        }
    }

.. code-block:: yaml

    App\Checker\ResourceChannelChecker:
        decorates: Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface
        arguments: ['@.inner']

In this moment access for our single resources should work properly with all restricties.

> For restricting our index, we need extend our resource repository for ``App\Repository\SupplierRepository`` with ``getQueryBuilder()`` method:

Reviewer 2: ```suggestion
To restrict our index, we need to extend our resource repository ``App\Repository\SupplierRepository`` with ``getQueryBuilder()`` method:
```

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Repository;

    use Doctrine\ORM\QueryBuilder;
    use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;

    final class SupplierRepository extends EntityRepository
    {
        public function getQueryBuilder(): QueryBuilder
        {
            return $this->createQueryBuilder('o');
        }
    }

> Remember to register ``App\Repository\SupplierRepository`` for resource:

Reviewer 2: ```suggestion
Remember to register ``App\Repository\SupplierRepository`` as a resource:
```

        {
            return ChoiceType::class;
        }

        public function getBlockPrefix(): string
        {
            return 'sylius_channel_choice';
        }
    }

.. code-block:: yaml

    App\Form\ChannelPerAdminChannelType:
        arguments: ['@sylius.repository.channel', '@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface']
        tags: ['form.type']

The ``Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface`` service allows getting a list of proper channels for the currently logged-in admin.

* Next, create a form type for your entity:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Form;

    use Sylius\Bundle\ResourceBundle\Form\Type\AbstractResourceType;
    use Symfony\Component\Form\Extension\Core\Type\TextType;
    use Symfony\Component\Form\FormBuilderInterface;

    final class SupplierType extends AbstractResourceType
    {
        public function buildForm(FormBuilderInterface $builder, array $options)
        {
            $builder
                ->add('name', TextType::class, [
                    'label' => 'Name'
                ])

>                 ->add('channel', ChannelPerAdminChannelType::class, [
>                     'label' => 'Channel'
>                 ])

Reviewer 4: Couldn't we just add here:
```suggestion
                ->add('channels', ChannelChoiceType::class, [
                    'choices' => $this->availableChannelsForAdminProvider->getChannels(),
                    'label' => 'sylius.ui.channel',
                ])
```
instead of adding the entire `ChannelPerAdminChannelType` form type?

            ;
        }

        public function getBlockPrefix(): string
        {
            return 'supplier';
        }
    }

.. code-block:: yaml

    App\Form\ChannelPerAdminChannelType:
            arguments: ['@sylius.repository.channel', '@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface']

>             tags: ['form.type']

Reviewer 4: Wrong indentation

Remember to register ``App\Form\ChannelPerAdminChannelType`` for resource:

.. code-block:: yaml

    sylius_resource:
    resources:
        app.supplier:
            driver: doctrine/orm
            classes:
                model: App\Entity\Supplier
                repository: App\Repository\SupplierRepository

>    +            form: App\Form\SupplierType

Reviewer 4: Wrong indentation

1. Add your entity to admin menu:
----------------------------------

Adding a new entity to the admin menu is described in the section ``How to customize Admin Menu`` of :doc:`this guide </customization/menu>`.

1. Restrict access to the entity for the respective channel administrator roles (using ACL/RBAC):

> -------------------------------------------------------------------------------------------------

Reviewer 4: Is this point necessary in cookbook for admin channel?

Author: IMO is better to have. But

Reviewer 5: IIRC channel admins and RBAC are different features, so we can mention RBAC but the configuration of the new RBAC role should be in a separate doc.

* Add access to all resource sections: ``index, create, update, delete, show, bulk_delete``

.. code-block:: yaml

    sylius_plus:
    permissions:
        app_admin_supplier_index:
            parent: suppliers
            label: action.index
        app_admin_supplier_create:
            parent: suppliers
            label: action.create
        app_admin_supplier_update:
            parent: suppliers
            label: action.update
        app_admin_supplier_delete:
            parent: suppliers
            label: action.delete
        app_admin_supplier_show:
            parent: suppliers
            label: action.show
        app_admin_supplier_bulk_delete:
            parent: suppliers

>             label: action.bulk_delete

Reviewer 4: Wrong indentation

.. note::

    More information about using administrator roles (ACL/RBAC) can be found :doc:`here </book/customers/admin_user>`.

* Create ``App\Checker\ResourceChannelEnabilibityChecker`` and decorate ``Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityCheckerInterface``.

* Then add ``Supplier`` as checking resource:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Checker;

    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityCheckerInterface;
    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityChecker as DecoratedResourceChannelEnabilibityChecker;

    final class ResourceChannelEnabilibityChecker implements ResourceChannelEnabilibityCheckerInterface
    {

>         /** @var DecoratedResourceChannelEnabilibityChecker */
>         private $decoratedResourceChannelEnabilibityChecker;
> 
>         public function __construct(DecoratedResourceChannelEnabilibityChecker $decoratedResourceChannelEnabilibityChecker)

Reviewer 4: ```suggestion
        /** @var ResourceChannelEnabilibityCheckerInterface */
        private $decoratedResourceChannelEnabilibityChecker;

        public function __construct(ResourceChannelEnabilibityCheckerInterface $decoratedResourceChannelEnabilibityChecker)
```

        {
            $this->decoratedResourceChannelEnabilibityChecker = $decoratedResourceChannelEnabilibityChecker;
        }

        public function forResourceName(string $resourceName): bool
        {

>             if (!$this->decoratedResourceChannelEnabilibityChecker->forResourceName($resourceName)) {
>                 return in_array($resourceName, [
>                     'supplier'
>                 ]);
>             }
> 
>             return true;

Reviewer 4: I looks clearer to me
```suggestion
            if ($this->decoratedResourceChannelEnabilibityChecker->forResourceName($resourceName)) {
                return true;
            }

            return in_array($resourceName, ['supplier']);
```

        }
    }

.. code-block:: yaml

    App\Checker\ResourceChannelEnabilibityChecker:
        decorates: Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityCheckerInterface
        arguments: ['@.inner']

* Create ``App\Checker\ResourceChannelChecker`` and decorate ``Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface`` next add condition for checking ``Supplier``.

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Checker;

    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface;
    use Sylius\Plus\Entity\ChannelInterface;

>     final class ResourceChannelChecker implements ResourceChannelCheckerInterface

Reviewer 4: We definitely should at least define in `ResourceChannelChecker`, in Plus the behaviour for checking the entity with `ChannelAwareInterface` and then recommend for implementing this interface instead of decorating this service

    {
        /** @var ResourceChannelCheckerInterface */
        private $decoratedResourceChannelChecker;

        public function __construct(ResourceChannelCheckerInterface $decoratedResourceChannelChecker)
        {
            $this->decoratedResourceChannelChecker = $decoratedResourceChannelChecker;
        }

        public function isFromChannel(object $resource, ChannelInterface $channel): bool
        {

>                 // this condition is needed because while creating a Symfony form (our edit/create section), the given object has all fields as null. After creating an empty form this iteration is repeated and properly checked.            if ($resource instanceof Supplier && $resource->getChannel() === null) {

Reviewer 4: ```suggestion
            // this condition is needed because while creating a Symfony form (our edit/create section), the given object has all fields as null. After creating an empty form this iteration is repeated and properly checked.            
            if ($resource instanceof Supplier && $resource->getChannel() === $channel) {
```

        decorates: Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface
        arguments: ['@.inner']

After that, access to the resource should work properly with all restrictions.

1. Extend the resource repository's method used for getting a list of Suppliers

* In order to view only Suppliers from the Channel to which the current administrator has access on the index of Suppliers, we need to extend the Supplier resource repository with ``getQueryBuilder()`` method:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Repository;

    use Doctrine\ORM\QueryBuilder;
    use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;

    final class SupplierRepository extends EntityRepository
    {
        public function getQueryBuilder(): QueryBuilder
        {
            return $this->createQueryBuilder('o');
        }
    }

* Remember to register ``App\Repository\SupplierRepository`` for resource:

.. code-block:: yaml

    sylius_resource:
    resources:
        app.supplier:
            driver: doctrine/orm
            classes:
                model: App\Entity\Supplier
                form: App\Form\SupplierType

>   +             repository: App\Repository\SupplierRepository

Reviewer 4: I'm almost sure that this part is not needed as the `getQueryBuilder` method exists in `EntityRepository`. Could you test that?

Author: `getQueryBuilder` doesn't exist but `createQueryBuilder` yes, I tried with it but I have problems.

            $this->supplierRepository = $supplierRepository;
        }

        public function create(): QueryBuilder
        {
            $listQueryBuilder = $this->supplierRepository->getQueryBuilder();

            /** @var ChannelInterface|null $channel */
            $channel = $this->adminChannelProvider->getChannel();
            if ($channel === null) {
                return $listQueryBuilder;
            }

            return $listQueryBuilder
                ->andWhere('o.channel = :channel')
                ->setParameter('channel', $channel)
            ;
        }
    }

.. code-block:: yaml

    App\Doctrine\ORM\RestrictingSupplierListQueryBuilder:
        public: true
        class: App\Doctrine\ORM\RestrictingSupplierListQueryBuilder
        arguments: ['@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface', '@app.repository.supplier']

* Add method to Suppliers grid:

.. code-block:: yaml

    sylius_grid:
        grids:
            app_admin_supplier:
                driver:
                    name: doctrine/orm
                    options:
                        class: App\Entity\Supplier
                        repository:

>   +                       method: [expr:service('App\\Doctrine\\ORM\\RestrictingSupplierListQueryBuilder'), create]

Reviewer 4: ```suggestion
  +                         method: [expr:service('App\\Doctrine\\ORM\\RestrictingSupplierListQueryBuilder'), create]
```
